tox_docker/tox4/plugin.py

```diff
@@ -0,0 +1,122 @@
+    EnvRunningContainers,
+    parse_container_config,
+)
+from tox_docker.tox4.log import log
+
+CONTAINER_CONFIGS: Dict[str, ContainerConfig] = {}
+ENV_CONTAINERS: EnvRunningContainers = {}
+
+
+@impl
+def tox_add_core_config(core_conf: ConfigSet, config: Config) -> None:
+    container_config_names = discover_container_configs(config)
+
+    # validate command line options
+    for container_name in config.options.docker_dont_stop:
+        if container_name not in container_config_names:
+            raise ValueError(
+                f"Container {container_name!r} not found (from --docker-dont-stop)"
+            )
+
+    for container_name in container_config_names:
+        CONTAINER_CONFIGS[container_name] = parse_container_config(
+            config, container_name, container_config_names
+        )
+
+
+@impl
+def tox_before_run_commands(tox_env: ToxEnv) -> None:
+    tox_env.conf.add_config(
+        keys=["docker"],
+        of_type=EnvList,
+        default=EnvList([]),
+        desc="docker image configs to load",
+        post_process=list,  # type: ignore
+    )
+
+    container_names = tox_env.conf["docker"]
+    env_container_configs = []
+
+    seen = set()
+    for container_name in container_names:
+        if container_name not in CONTAINER_CONFIGS:
+            raise ValueError(f"Missing [docker:{container_name}] in tox.ini")
+        if container_name in seen:
+            raise ValueError(f"Container {container_name!r} specified more than once")
+        seen.add(container_name)
+        env_container_configs.append(CONTAINER_CONFIGS[container_name])
+
+    for container_config in env_container_configs:
+        docker_pull(container_config, log)
+
+    ENV_CONTAINERS.setdefault(tox_env, {})
+    containers: RunningContainers = ENV_CONTAINERS[tox_env]
+
+    for container_config in env_container_configs:
+        container = docker_run(container_config, containers, log)
+        containers[container_config.name] = container
+
+    for container_name, container in containers.items():
+        container_config = CONTAINER_CONFIGS[container_name]
+        try:
+            docker_health_check(container_config, container, log)
+        except HealthCheckFailed:
+            tox_env.interrupt()
+            clean_up_containers(tox_env)
+            raise
+
+    for container_name, container in containers.items():
+        container_config = CONTAINER_CONFIGS[container_name]
+        tox_env.conf["set_env"].update(get_env_vars(container_config, container))
+
+
+@impl
+def tox_after_run_commands(
+    tox_env: ToxEnv, exit_code: int, outcomes: List[Outcome]
+) -> None:
+    clean_up_containers(tox_env)
+
+
+def clean_up_containers(tox_env: ToxEnv) -> None:
+    env_containers: RunningContainers = ENV_CONTAINERS.get(tox_env, {})
+    containers_and_configs = [
+        (CONTAINER_CONFIGS[name], container)
+        for name, container in env_containers.items()
+    ]
+    stop_containers(containers_and_configs, log)
```

Comment:
This probably should be something like:

```
for container in tox_env.conf['docker_container']:
    container.stop()
```

Reply:
It may not be worth the trouble, but my thought was to have a "business logic" definition in `tox_docker/plugin.py`, and then in the tox3 and tox4 versions to have adapters to the tox version. The "business logic" in question here is checking the `stop` flag (based on a command-line option, but propagated through to the `ContainerConfig` class), so admittedly it's maybe a weak case for DRY, but that's what I was after.

Reply:
I guess. Though I'm surprised you want to support both tox 3 and 4... I'd personally drop tox 3 support with the move.